cognite/client/utils/_concurrency.py

```diff
@@ -109,38 +110,109 @@ def collect_exc_info_and_raise(
         ) from dup_exc
 
 
-def execute_tasks_concurrently(
-    func: Callable, tasks: Union[Sequence[Tuple], List[Dict]], max_workers: int
-) -> TasksSummary:
+T_Result = TypeVar("T_Result", covariant=True)
+
+
+class Executor(Protocol):
+    def submit(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any) -> Future[T_Result]:
+        ...
+
+
+class Future(Protocol[T_Result]):
+    def result(self) -> T_Result:
+        ...
+
+
+class SyncFuture(Future):
+    def __init__(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any):
+        self.__fn = fn
+        self.__args = args
+        self.__kwargs = kwargs
+
+    def result(self) -> T_Result:
+        return self.__fn(*self.__args, **self.__kwargs)
```

Comment:
Why not use _WorkItem from threading?

https://github.com/python/cpython/blob/c2c970fc26e6de5b9e338413eac71d46aea5ca5e/Lib/concurrent/futures/thread.py#L46

Reply:
Don't like using protected code

Reply:
Fair enough! But we shouldn't (probably) overload typical class names from `threading` like `Future`, likely to become an import in a utils module named "concurrency" at some point.

Reply:
agree. Will rename Future and Executor to _Future and _Executor.